### ISSUE TITLE
CFFI void* in nopython mode, determine Numba type of `_cffi_backend.CData` 

### ISSUE THREAD

**Issue by jpata** [numba/numba#4115]
## Feature request

tl;dr: support for `void*` as a nopython mode argument.

Numba can make use of CFFI in nopython mode to call simple C functions with little overhead. More complex functions might rely on state, in particular, passing structured data as a `void*`. Currently I could not find a straightforward way of doing that with Numba.

A practical example might be dealing with a C wrapper of a C++ code such as Cairo, where `void*` are used to pass around the underlying class. I have created a minimal standalone example below, consisting of a C++ library from `mylib.cc` and a python code `test.py`.

## Minimal example

mylib.cc:
~~~C++
//compile with  g++ -fPIC -shared -o mylib.dylib mylib.cc
class MyClass {
public:
  
  MyClass(const char* filename) {
    //init class, e.g. load files
  }
  
  double func(int a) {
    //an expensive C++ function
    return 0.0;
  }
  
};

extern "C" {
  void* new_class(const char* filename) {
    return new MyClass(filename);
  }

  double class_func(void* cls, int a) {
    return ((MyClass*)cls)->func(a);
  }
}

~~~

test.py:
~~~python
import numba
from cffi import FFI

ffi = FFI()
ffi.cdef("""
void* new_class(const char* filename);
double class_func(void* cls, int a);
""")
mylib = ffi.dlopen("mylib.dylib")

class_func = mylib.class_func

c = mylib.new_class("asd")

@numba.njit
def call_myfunc(a):
    x = class_func(c, a)

#throws exception as the type of c ('_cffi_backend.CData') is not known to Numba
call_myfunc(1)
~~~

This currently fails with
~~~
cannot determine Numba type of <class '_cffi_backend.CData'>
~~~
as the type of `c` cannot be propagated in the nopython mode. With `nopython=False`, this works, but it would be great not to have to pay the additional overhead :)

## Who would benefit from it? 
  - In my case, I'm trying to wrap some existing scientific C++ code that relies on classes that can be easily encapsulated in a few C functions and passing around a blind pointer to the class.
  - In gitter, @tillsten mentioned that he had the same issue when wrapping Cairo.
  - I have found a work-around for this [here](https://github.com/willsheffler/worms/blob/1d31a23890b059baccf062ace98c7ec63cd61554/worms/khash/khash_cffi.py#L43), where it seems `register_type` did not give the expected result, but `register_module` worked. I was not successful in using this approach for my purpose.

cc @stuartarchibald, who proposed a solution based on adding additional intrinsics: [May 23, 2019 1:08 PM](https://gitter.im/numba/numba?at=5ce6fdcf83ae782aeebbf941)

**Comment by stuartarchibald** [comment]
@jpata Thanks for writing this, we'll take a look.

**Comment by sklam** [comment]
The `register_type` and `register_module` approach requires an [out-of-line CFFI module](https://cffi.readthedocs.io/en/latest/overview.html#purely-for-performance-api-level-out-of-line), which doesn't apply in this case due to the use of C++.

After discussing with fellow numba devs, we agree that the ability to use CFFI `void*` data will be useful.  However, these CFFI data must be passed as arguments and cannot be of global variables, which can lead to various safety concerns.

It will be a low-hanging fruit if we focus on support `cffi.CData` of `void*` such that they can be used as function arguments as well as regular value inside a jit function.

**Comment by slayoo** [comment]
Am I right that this blocks the following code from working (starting point: https://github.com/mpi4py/mpi4py/blob/master/demo/wrap-cffi/helloworld.py): 

```python
import numba
import cffi
import platform
import numpy as np
from mpi4py import MPI


ffi = cffi.FFI()

if MPI._sizeof(MPI.Comm) == ffi.sizeof('int'):
    _mpi_comm_t = 'int'
else:
    _mpi_comm_t = 'void*'
ffi.cdef(f"typedef {_mpi_comm_t} MPI_Comm;")
ffi.cdef("int MPI_Initialized(int *flag);")
ffi.cdef("int MPI_Comm_size(MPI_Comm comm, int *size);")

sys = platform.system()

if sys == 'Linux':
    libmpi = ffi.dlopen('libmpi.so')
else:
    raise NotImplementedError()

_MPI_Comm_size = libmpi.MPI_Comm_size
_MPI_Initialized = libmpi.MPI_Initialized

comm_ptr = MPI._addressof(MPI.COMM_WORLD)
comm_val = ffi.cast('MPI_Comm*', comm_ptr)[0]


@numba.njit()
def initialized():
    flag = np.empty((1,), dtype=np.int32)
    status = _MPI_Initialized(ffi.from_buffer(flag))
    assert status == 0
    return bool(flag[0])


@numba.njit()
def size():
    value = np.empty(1, dtype=np.int32)
    status = _MPI_Comm_size(comm_val, ffi.from_buffer(value))
    assert status == 0
    return value[0]


if __name__ == '__main__':
    assert initialized()  # works!
    print(size.py_func())  # works!
    assert size.py_func() == size()  # njitted version fails
```

I.e., size.py_func() works, while size() fails with:
```
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'comm_val': cannot determine Numba type of <class '_cffi_backend.CData'>
```

I believe, this would be another good use case, and does not relate to C++.
Of course, if there is any way around that could make this work, help very welcome!
Thanks
